cloud/db_cloud_test.cc

```diff
@@ -1884,26 +1884,53 @@ TEST_F(CloudTest, LiveFilesConsistentAfterApplyLocalCloudManifestDeltaTest) {
 TEST_F(CloudTest, WriteAfterUpdateCloudManifestArePersistedInNewEpoch) {
   cloud_env_options_.cookie_on_open = "1";
   OpenDB();
-  ASSERT_OK(db_->Put(WriteOptions(), "Hello", "world"));
+  ASSERT_OK(db_->Put(WriteOptions(), "Hello1", "world1"));
   ASSERT_OK(db_->Flush(FlushOptions()));
 
   std::string new_cookie = "2";
   std::string new_epoch = "dca7f3e19212c4b3";
   ASSERT_OK(GetCloudEnvImpl()->ApplyLocalCloudManifestDelta(
       dbname_, new_cookie,
       CloudManifestDelta{GetDBImpl()->TEST_Current_Next_FileNo(), new_epoch}));
+  ASSERT_OK(GetCloudEnvImpl()->UploadLocalCloudManifest(dbname_, new_cookie));
```

Comment:
If the process dies here the cloud state will be inconsistent. The CLOUDMANIFEST's current epoch will not point to any MANIFEST file, because it will only be created after the next manifest write.

If we are going to keep the epoch suffix in local MANIFEST file the sequence of steps on the epoch roll should be:
1. Either copy MANIFEST-oldepoch into MANIFEST-newepoch (both locally and in the cloud) or create a MANIFEST snapshot into MANIFEST-newepoch (both locally and in the cloud).
2. Create new CLOUDMANIFEST with the new epoch and upload it to the cloud.
3. Delete MANIFEST-oldepoch.

We obviously need to ensure that no MANIFEST writes happen from 1-3 (we can assume that, as this is happening on the follower) and that new MANIFEST writes after (3) go to the MANIFEST-newepoch.

That way the state both locally and in the cloud is in the consistent state at all stages. If the process dies before (2) we'll still be in the old epoch and if the process dies after (2) we'll be in the new epoch.

We need to think more about what happens if local CLOUDMANIFEST has the new epoch, while the cloud manifest doesn't (i.e. the process failed before we managed to upload CM To the cloud).

Reply:
```
If the process dies here the cloud state will be inconsistent. The CLOUDMANIFEST's current epoch will not point to any MANIFEST file
```
It wouldn't actually. In `UploadLocalCloudManifest`, we uploaded both MANFIEST file and CLOUDMANIFEST to s3.

If process dies in the middle, in the worst case, MANIFEST file with new epoch exists but not CLOUDMANIFEST, which is ok.

I should add more tests on this part.

Reply:
```
We obviously need to ensure that no MANIFEST writes happen from 1-3 (we can assume that, as this is happening on the follower) and that new MANIFEST writes after (3) go to the MANIFEST-newepoch
```

Yes. We have to ensure there is no MANIFEST writes on follower when switching CM/M. We will shutdown the `LeafFollowerUpdater` during leader election.

Reply:
> It wouldn't actually. In UploadLocalCloudManifest, we uploaded both MANFIEST file and CLOUDMANIFEST to s3.
> If process dies in the middle, in the worst case, MANIFEST file with new epoch exists but not CLOUDMANIFEST (since we upload MANIFEST file first), which is ok

Good point. How about local state, assuming resync_on_open=false?

Reply:
When `ApplyLocalCloudManifestDelta`, we generate new local `MANIFEST-epoch` first and then new local `CLOUDMANIFEST-cookie`.  So we maintain the invariant that: `as long as CLOUDMANIFEST-cookie exists locally, it will have a corresponding MANIFEST-epoch locally`

Also, if process dies in the middle, it means the follower fails to take over. If it wants to start another round of taking over, it will have to:
1. reopen DB with existing cookie in FDB (which != the cookie used in previous round of taking over)
2. CAS(new_leader, new_epoch), and call `ApplyLocalCloudManifestDelta(new_epoch)`. This new_epoch != the cookie used in previous round of taking over as well.

Reply:
Sounds good. I left some comments on the internal design doc, let's continue the conversation there. We can ship this PR, though.